Initial ICA integration, missing auth setup

ethanfrey opened the pull request:
Based on [`gaiad` integration](https://github.com/cosmos/gaia/blob/v7.0.1/app/app.go) (for host) and referencing #793 for the ice controller setup. 

Closes #806

Replaces #793

This enables ICA on the default wasmd application.

However, it is very unclear how to add proper icaController support

ethanfrey commented on app/app.go:
```diff
@@ -434,32 +446,41 @@ func NewWasmApp(
 
-	// icaControllerIBCModule := icacontroller.NewIBCModule(app.icaControllerKeeper, icaAuthModule)
-	// icaHostIBCModule := icahost.NewIBCModule(app.icaHostKeeper)
+	// TODO: clarify how to build icaAuthModule or remove icaController support?
```
I am a bit lost here.
Can anyone show me an example of non-mock authorization for an icaController?

Otherwise, I would just make this a host... I want to integrate tried-and-true ibc code, not experiment here.

alpe replied:
Just host makes sense for me, in this PR

ethanfrey commented:
Happy for a review here.

I could also merge as is (with icaController not working) for the demo of ica integration, and add #839 later, to allow icaController once the downstream dependencies get worked out.

alpe commented on app/app.go:
```diff
@@ -290,7 +300,7 @@ func NewWasmApp(
 		govtypes.StoreKey, paramstypes.StoreKey, ibchost.StoreKey, upgradetypes.StoreKey,
 		evidencetypes.StoreKey, ibctransfertypes.StoreKey, capabilitytypes.StoreKey,
-		feegrant.StoreKey, authzkeeper.StoreKey, wasm.StoreKey,
+		feegrant.StoreKey, authzkeeper.StoreKey, wasm.StoreKey, icahosttypes.StoreKey, icacontrollertypes.StoreKey,
```
`icacontrollertypes` not required for host role

alpe commented on app/app.go:
```diff
@@ -74,6 +74,14 @@ import (
+	ica "github.com/cosmos/ibc-go/v3/modules/apps/27-interchain-accounts"
+	icacontroller "github.com/cosmos/ibc-go/v3/modules/apps/27-interchain-accounts/controller"
+	icacontrollerkeeper "github.com/cosmos/ibc-go/v3/modules/apps/27-interchain-accounts/controller/keeper"
+	icacontrollertypes "github.com/cosmos/ibc-go/v3/modules/apps/27-interchain-accounts/controller/types"
```
`icacontrollertypes` is not required for host role

alpe commented on app/app.go:
```diff
@@ -74,6 +74,14 @@ import (
 	upgradekeeper "github.com/cosmos/cosmos-sdk/x/upgrade/keeper"
 	upgradetypes "github.com/cosmos/cosmos-sdk/x/upgrade/types"
+	ica "github.com/cosmos/ibc-go/v3/modules/apps/27-interchain-accounts"
+	icacontroller "github.com/cosmos/ibc-go/v3/modules/apps/27-interchain-accounts/controller"
```
`icacontroller` is not required for the host role

alpe commented on app/app.go:
```diff
@@ -323,6 +333,8 @@ func NewWasmApp(
 	)
 	scopedIBCKeeper := app.capabilityKeeper.ScopeToModule(ibchost.ModuleName)
+	scopedICAHostKeeper := app.capabilityKeeper.ScopeToModule(icahosttypes.SubModuleName)
+	scopedICAControllerKeeper := app.capabilityKeeper.ScopeToModule(icacontrollertypes.SubModuleName)
```
not required for host role

alpe commented on app/app.go:
```diff
@@ -228,29 +236,31 @@ type WasmApp struct {
+	paramsKeeper        paramskeeper.Keeper
+	evidenceKeeper      evidencekeeper.Keeper
+	ibcKeeper           *ibckeeper.Keeper // IBC Keeper must be a pointer in the app, so we can SetRouter on it correctly
+	icaControllerKeeper icacontrollerkeeper.Keeper
```
not required for host role

alpe commented on app/app.go:
```diff
@@ -434,32 +446,41 @@ func NewWasmApp(
-	// 	app.ibcKeeper.ChannelKeeper, &app.ibcKeeper.PortKeeper,
-	// 	scopedicaControllerKeeper, app.MsgServiceRouter(),
-	// )
+	_ = app.getSubspace(icahosttypes.SubModuleName)
```
nit: please remove

alpe commented on app/app.go:
```diff
@@ -434,32 +446,41 @@ func NewWasmApp(
+		scopedICAHostKeeper,
+		app.MsgServiceRouter(),
+	)
+	app.icaControllerKeeper = icacontrollerkeeper.NewKeeper(
```
can be removed for host role

alpe commented on app/app.go:
```diff
@@ -434,32 +446,41 @@ func NewWasmApp(
+		scopedICAControllerKeeper,
+		app.MsgServiceRouter(),
+	)
+	icaModule := ica.NewAppModule(&app.icaControllerKeeper, &app.icaHostKeeper)
```
```suggestion
	icaModule := ica.NewAppModule(nil, &app.icaHostKeeper)
```

alpe commented on app/app.go:
```diff
@@ -434,32 +446,41 @@ func NewWasmApp(
+	icaHostIBCModule := icahost.NewIBCModule(app.icaHostKeeper)
 
-	// // initialize ICA module with mock module as the authentication module on the controller side
+	// FIXME: this came from PR #793, but I don't like testing "mocks" in a real app
```
please remove comments and code example. We can always add this later when required for a concrete test

alpe commented on app/app.go:
```diff
@@ -434,32 +446,41 @@ func NewWasmApp(
+	// See https://github.com/cosmos/ibc-go/blob/v3.0.0/docs/apps/interchain-accounts/integration.md
+	// "Note: No `icaauth` exists, this must be substituted with an actual Interchain Accounts authentication module"
+	var icaAuthModule porttypes.IBCModule
+	icaControllerIBCModule := icacontroller.NewIBCModule(app.icaControllerKeeper, icaAuthModule)
```
let's not have ica controller setup here

alpe commented on app/app.go:
```diff
@@ -718,6 +743,8 @@ func NewWasmApp(
 	app.scopedTransferKeeper = scopedTransferKeeper
 	app.scopedWasmKeeper = scopedWasmKeeper
+	app.scopedICAHostKeeper = scopedICAHostKeeper
+	app.scopedICAControllerKeeper = scopedICAControllerKeeper
```
no controller needed

alpe commented on app/app.go:
```diff
@@ -865,6 +892,8 @@ func initParamsKeeper(appCodec codec.BinaryCodec, legacyAmino *codec.LegacyAmino
 	paramsKeeper.Subspace(ibctransfertypes.ModuleName)
 	paramsKeeper.Subspace(ibchost.ModuleName)
+	paramsKeeper.Subspace(icahosttypes.SubModuleName)
+	paramsKeeper.Subspace(icacontrollertypes.SubModuleName)
```
no controller needed

alpe commented on app/app.go:
```diff
@@ -508,11 +529,11 @@ func NewWasmApp(
-		AddRoute(ibctransfertypes.ModuleName, transferIBCModule)
+		AddRoute(ibctransfertypes.ModuleName, transferIBCModule).
+		AddRoute(icacontrollertypes.SubModuleName, icaControllerIBCModule).
+		AddRoute(icahosttypes.SubModuleName, icaHostIBCModule)
```
👍

alpe reviewed (approved):
Looks complete. I have added a lot of comments about the ICA controller setup.
If you want to be strict on the host role, you can remove some setup

ethanfrey commented:
Removed all controller code in https://github.com/CosmWasm/wasmd/pull/837/commits/38932b7149f58518e3bf519e358f2dd7f673a6dd as we only use host mode here

Added it back in #839 as https://github.com/CosmWasm/wasmd/pull/839/commits/9b8fb98acb592f0d66ff3556bbf78379dc1d7b18 which can be used as a good start for any real controller+auth integration
